client/blocks/reader-site-notification-settings/index.jsx

```diff
@@ -37,9 +37,15 @@ class ReaderSiteNotificationSettings extends Component {
 	iconRef = createRef();
 	spanRef = createRef();
 
-	UNSAFE_componentWillReceiveProps( nextProps ) {
-		if ( nextProps.emailDeliveryFrequency !== this.props.emailDeliveryFrequency ) {
-			this.setState( { selected: nextProps.emailDeliveryFrequency } );
+	updateSelectedState = ( selected ) => {
+		this.setState( ( prevState ) => {
+			return prevState.selected !== selected ? { selected } : null;
+		} );
+	};
+
+	componentDidUpdate( prevProps ) {
+		if ( this.props.emailDeliveryFrequency !== prevProps.emailDeliveryFrequency ) {
+			this.updateSelectedState( this.props.emailDeliveryFrequency );
```

Comment:
I'm not really happy with this kind of solution, there has to be a better one 😄 – I see this pattern quite often throughout Calypso where a form component is "reset" after a prop changes to make sure local state is in sync with whatever it should be according to the server.

Intentionally I would just call `setState` conditionally in `cDU` without extracting it to `updateSelectedState`. However, that triggers a warning about calling `setState` in `cDU` (legitimately imo) so I just "hide" that `setState` call from eslint which kinda smells.

Reply:
If there is some condition where the component should reset and forget all previous state, it means the same thing as unmounting the component and mounting it again, with a new fresh instance.

You can do that reset with `getDerivedStateFromProps`, but that's difficult to implement and it's easy to make a mistake (detecting the condition change incorrectly, resetting only part of the state, ...). And implementing that with hooks is even more tricky.

The easier way to reset (i.e., remount) a component on certain condition is to put that condition into a `key`:
```js
<TermFormDialog key={ term } term={ term } />
```
The `key` can be anything, not just a string. How it works is that when the new `key` is not `===` to the old key, the component gets remounted.

The downside of `key` is that the component user must not forget to add it. In our case, it could be ensured by doing a wrapper component:
```js
function TermFormDialogInner( props ) {
  /* A big function body with all the dialog stuff */
}

export function TermFormDialog( props ) {
  return <TermFormDialogInner { ...props } key={ props.term } />
}
```

Another solution that works well for this one specific `TermFormDialog` component, which is not widely reused, is to ensure that none of its usages ever change `term` in mid-flight, and then just ignore the problem. The component won't be 100% resilient, but in our case, after we refactor away the `showDialog` prop, it will be OK.

Reply:
Seems to me that the `selected` state doesn't need to be there at all. The state in Redux, selected into the `emailDeliveryFrequency` and updated by the `updateNewPostEmailSubscription` action, is always correct.

The `updateNewPostEmailSubscription` action synchronously updates the Redux state, so that `emailDeliveryFrequency` immediately changes and the correct item is selected in the UI. Then the action triggers a REST request, and either confirms or reverts the changes when the request finishes. In other words, `emailDeliveryFrequency` is optimistically updated and reverted on failure.

Reply:
You're right 👍🏻 https://github.com/Automattic/wp-calypso/pull/57987/commits/5af870cb8c7a305792b1c811d1c6803651ea4bfa